File: src/investraton/engine/plan_align.py
"""Plan-alignment: compare the portfolio's actual shape to the user's stated plan.

This is the differentiator from the original design — it lets the system say
"your growth bucket is running hot vs your 15% target" or "semis weakness lands in
a bucket you're underweight", purely as *alignment detection*. Never advice.
"""

from __future__ import annotations

from collections import defaultdict

from ..config import Plan
from ..models import Finding, Holding

DRIFT_NOTE = 0.05   # 5 percentage points
DRIFT_WATCH = 0.10  # 10 percentage points

# Friendly labels for plan buckets / holding types.
_LABELS = {
    "etf": "ETFs",
    "growth_stock": "Growth stocks",
    "experimental": "Experimental",
    "equity": "Other equities",
}


def _label(key: str) -> str:
    return _LABELS.get(key, key.replace("_", " ").title())


def analyse_plan(holdings: list[Holding], plan: Plan) -> list[Finding]:
    targets = plan.allocation_targets or {}
    if not targets or not holdings:
        return []

    current: dict[str, float] = defaultdict(float)
    for h in holdings:
        current[h.type] += h.portfolio_weight

    findings: list[Finding] = []
    buckets = set(targets) | set(current)
    drifts = []
    for b in buckets:
        cur = current.get(b, 0.0)
        tgt = targets.get(b, 0.0)
        drifts.append((b, cur, tgt, cur - tgt))

    # Headline: overall alignment summary across known plan buckets.
    summary = "; ".join(
        f"{_label(b)} {current.get(b,0)*100:.0f}% (target {targets[b]*100:.0f}%)"
        for b in targets
    )
    findings.append(Finding(
        kind="plan_drift",
        title="Allocation vs plan",
        detail=f"Current vs target: {summary}.",
        severity="info",
    ))

    # Individual drifts worth flagging.
    for b, cur, tgt, drift in sorted(drifts, key=lambda x: abs(x[3]), reverse=True):
        if b not in targets:
            if cur >= DRIFT_NOTE:
                findings.append(Finding(
                    kind="plan_drift",
                    title=f"{_label(b)} sits outside your plan ({cur*100:.0f}%)",
                    detail=f"{cur*100:.0f}% of your portfolio is in '{_label(b)}', which isn't a bucket in your "
                           f"allocation plan. Either it's intentional or your plan needs updating.",
                    severity="note",
                ))
            continue
        if abs(drift) < DRIFT_NOTE:
            continue
        over = drift > 0
        findings.append(Finding(
            kind="plan_drift",
            title=f"{_label(b)} is {'over' if over else 'under'} target by {abs(drift)*100:.0f}pp",
            detail=f"{_label(b)} is {cur*100:.0f}% vs your {tgt*100:.0f}% target "
                   f"({'+' if over else ''}{drift*100:.0f}pp). "
                   + (f"Your €{plan.monthly_contribution_eur:,.0f}/month could be steered toward "
                      f"under-weight buckets if you want to track the plan."
                      if not over and plan.monthly_contribution_eur else
                      "Worth knowing when you deploy your next contribution."),
            severity="watch" if abs(drift) >= DRIFT_WATCH else "note",
        ))
    return findings

File: engine/src/planAlign.ts
/** Port of engine/plan_align.py — actual allocation vs the user's stated plan.
 *
 * Alignment detection only; never advice.
 */

import type { Finding, Holding, Plan } from "./models.ts";
import { pyFormat, pyThousands } from "./round.ts";

const DRIFT_NOTE = 0.05;  // 5 percentage points
const DRIFT_WATCH = 0.10; // 10 percentage points

const LABELS: Record<string, string> = {
  etf: "ETFs",
  growth_stock: "Growth stocks",
  experimental: "Experimental",
  equity: "Other equities",
};

function label(key: string): string {
  if (key in LABELS) return LABELS[key];
  // Python: key.replace("_", " ").title()
  return key.replaceAll("_", " ").replace(/\w\S*/g, (w) =>
    w.charAt(0).toUpperCase() + w.slice(1).toLowerCase());
}

const f0 = (x: number): string => pyFormat(x, 0);
const fThousands = (x: number): string => pyThousands(x);

export function analysePlan(holdings: Holding[], plan: Plan): Finding[] {
  const targets = plan.allocation_targets ?? {};
  if (Object.keys(targets).length === 0 || holdings.length === 0) return [];

  const current: Record<string, number> = {};
  for (const h of holdings) current[h.type] = (current[h.type] ?? 0) + h.portfolio_weight;

  const findings: Finding[] = [];
  const buckets = new Set([...Object.keys(targets), ...Object.keys(current)]);
  const drifts: Array<[string, number, number, number]> = [];
  for (const b of buckets) {
    const cur = current[b] ?? 0;
    const tgt = targets[b] ?? 0;
    drifts.push([b, cur, tgt, cur - tgt]);
  }

  // Headline: overall alignment across known plan buckets.
  const summary = Object.keys(targets)
    .map((b) => `${label(b)} ${f0((current[b] ?? 0) * 100)}% (target ${f0(targets[b] * 100)}%)`)
    .join("; ");
  findings.push({
    kind: "plan_drift",
    title: "Allocation vs plan",
    detail: `Current vs target: ${summary}.`,
    severity: "info",
  });

  // Individual drifts worth flagging.
  drifts.sort((a, b) => Math.abs(b[3]) - Math.abs(a[3]));
  for (const [b, cur, tgt, drift] of drifts) {
    if (!(b in targets)) {
      if (cur >= DRIFT_NOTE) {
        findings.push({
          kind: "plan_drift",
          title: `${label(b)} sits outside your plan (${f0(cur * 100)}%)`,
          detail: `${f0(cur * 100)}% of your portfolio is in '${label(b)}', which isn't a bucket in ` +
            `your allocation plan. Either it's intentional or your plan needs updating.`,
          severity: "note",
        });
      }
      continue;
    }
    if (Math.abs(drift) < DRIFT_NOTE) continue;
    const over = drift > 0;
    const tail = !over && plan.monthly_contribution_eur
      ? `Your €${fThousands(plan.monthly_contribution_eur)}/month could be steered toward ` +
        `under-weight buckets if you want to track the plan.`
      : "Worth knowing when you deploy your next contribution.";
    findings.push({
      kind: "plan_drift",
      title: `${label(b)} is ${over ? "over" : "under"} target by ${f0(Math.abs(drift) * 100)}pp`,
      detail: `${label(b)} is ${f0(cur * 100)}% vs your ${f0(tgt * 100)}% target ` +
        `(${over ? "+" : ""}${f0(drift * 100)}pp). ` + tail,
      severity: Math.abs(drift) >= DRIFT_WATCH ? "watch" : "note",
    });
  }
  return findings;
}
